cmake: fix conversion of zero to hex

attie-argentum opened the pull request:
Previously, `to_hex(0 output)` would have placed `0x` into the `output` variable, which is undesirable... this fix ensures that `0x0` is placed into this variable if the input is zero.

attie-argentum committed:
cmake: hex: fix conversion of zero to hex

Previously, `to_hex(0 output)` would have placed `0x` into the `output`
variable, which is undesirable... this fix ensures that `0x0` is placed
into this variable if the input is zero.

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

57300 commented:
No objections, but note that `from_hex()` and `to_hex()` were imported 7 years ago, and they can now be replaced with:
```cmake
math(EXPR dec "${hex}" OUTPUT_FORMAT DECIMAL)
math(EXPR hex "${dec}" OUTPUT_FORMAT HEXADECIMAL)
```
The only difference is that the output `hex` here is lowercase. Other than that, we can probably get rid of `hex.cmake`.

attie-argentum commented:
I don't believe `from_hex()` is actually used anywhere, and `to_hex()` is only used in this instance, so I'll rework the PR - thanks!

attie-argentum commented:
@57300 - thanks for your comment... I'm done!

57300 reviewed (dismissed):
Thanks! LGTM

tejlmand commented on cmake/hex.cmake:
```diff
@@ -1,56 +0,0 @@
-    set(${DEC} ${_res} PARENT_SCOPE)
-endfunction()
-
-function(to_hex DEC HEX)
```
no problem in changing the implementation of this, but we should deprecate the function before removing it.

There might be out-of-tree users relying on this function.

Please keep the function, and make an update in `FindDeprecated()` so we remember to remove the function after two releases:
https://github.com/zephyrproject-rtos/zephyr/blob/9074feabf6f827f6b37a69d77a5e13093799a905/cmake/modules/FindDeprecated.cmake#L5-L16

attie-argentum replied:
Thanks for the comment - I'll get on this, hopefully later today.

attie-argentum replied:
I'm not 100% clear on how to action this... Obviously we'd like a warning printed when anyone uses the `to_hex()` and `from_hex()` functions, but moving them to `cmake/modules/FindDeprecated.cmake` will make them "_disappear_" for anyone currently using them.

Additionally, we can't easily use the `find_package(Deprecated COMPONENTS ...)` used elsewhere, because `cmake/modules/version.cmake` seems to be included very early on with `CMAKE_MODULE_PATH` unset (and thus `.../FindDeprecated.cmake` is not found).

I'm going to propose a `message(DEPRECATION ...)` in `cmake/hex.cmake`, which is presented when it's included. I hope that suffices.

tejlmand replied:
> because cmake/modules/version.cmake seems to be included very early on with CMAKE_MODULE_PATH unset (and thus .../FindDeprecated.cmake is not found).

but it's not the `version.cmake` which should call `FindDeprecated`, it should be `hex.cmake` itself.
`version.cmake` itself has already converted to use `math()`, but you do have a point in `CMAKE_MODULE_PATH` not being set in all cases, and although `hex.cmake` itself could update `CMAKE_MODULE_PATH` when called, then we better avoid it, especially as this is used to determine correct Zephyr to use, based on versions.

> I'm going to propose a message(DEPRECATION ...) in cmake/hex.cmake, which is presented when it's included. I hope that suffices.

I believe that to be a good alternative in this case.

tejlmand reviewed (changes requested):
thanks for cleaning up this, and good to switch to CMake built-in.
Original code is from before CMake 3.13.

But we should deprecate the functions first, before removing them, as there can be oot users relying on them.

attie-argentum committed:
cmake: hex: deprecate use of to_hex() and from_hex() utility functions

Since v3.13, CMake has supported math(... OUTPUT_FORMAT <format>) and 0x
prefixes on numbers, which together provides the same functionality as
to_hex() and from_hex() that we've previously maintained.

Users should switch to using math() instead, and our functions are now
marked as deprecated.

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

attie-argentum committed:
cmake: version: stop using to_hex() in favour of CMake's math()

Since v3.13, CMake has supported math(... OUTPUT_FORMAT <format>), which
will perform the same functionality as to_hex().

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

attie-argentum commented:
In light of @tejlmand's comments, I've reworked this to:

1. Fix the bug
2. Mark the `to_hex()` and `from_hex()` functions as deprecated, and include a "_stub_" / reminder in `.../FindDeprecated.cmake`
3. Stop using `to_hex()` in `.../version.cmake`

I hope this works for everyone!

tejlmand commented on cmake/modules/FindDeprecated.cmake:
```diff
@@ -99,6 +99,13 @@ if("PYTHON_PREFER" IN_LIST Deprecated_FIND_COMPONENTS)
   endif()
 endif()
 
+if("HEX_CONV" IN_LIST Deprecated_FIND_COMPONENTS)
+  # This code was deprecated after Zephyr v3.5.0
+  # The deprecated warning is produced by hex.cmake itself
+  list(REMOVE_ITEM Deprecated_FIND_COMPONENTS HEX_CONV)
+  include(${ZEPHYR_BASE}/cmake/hex.cmake)
+endif()
+
```
This code doesn't make much sense now, as no one will be calling `find_package(Deprecated COMPONENTS HEX_CONV)`

```suggestion
```

attie-argentum replied:
I'm not sure how things marked for deprecation are tracked... as I understood it, CMake components that are "_to be removed_" need to be listed here, so it's there as a reminder more than anything else.

I'll happily remove if you're okay with that.

attie-argentum replied:
Ok, done - thanks!

tejlmand reviewed (changes requested):
looks good, only a minor nit.

Going to perform a couple of extra tests with multiple (and also older) Zephyr installations, just to ensure we don't hit any surprises in the lookup mechanism of Zephyr.
